### PR TITLE
Doio kb38

### PR DESCRIPTION
**Pull request by PepperKats**
## Description

I've added support for the Doio kb38. This is my first time with a pull request, so sorry if I didn't do it right. I've been using this for a while, so it seems to work pretty well.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb38/rules.mk [hunk 1]
```diff
@@ -0,0 +1,4 @@
+RGB_MATRIX_CUSTOM_KB = yes
+OLED_ENABLE = yes
+ENCODER_ENABLE = yes
```
[the remark this answers is not in the document]
Comment (PepperKats) [reply keyboards/doio/kb38/rules.mk:3]: I can't seem to find any references to doing that in the docs. Do you have a link I could look at?

File: keyboards/doio/kb38/kb38.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7, 0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF, 0xD0, 0xD1, 0xD2, 0xD3, 0xD4, 0x00
+    };
+
+    oled_write_P(qmk_logo, false);
+}
+
+bool oled_task_user(void) {
+    render_logo();
+    return false;
+}
```
Comment (waffle87) [inline keyboards/doio/kb38/kb38.c:33]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) { return false; }
    render_logo();
    return true;
}
```

File: keyboards/doio/kb38/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+#define _QWERTY 0
+#define _LAYERTWO 1
```
Comment (waffle87) [inline keyboards/doio/kb38/keymaps/default/keymap.c:20]:
```suggestion
enum layers {
    _QWERTY,
    _LAYERTWO
};
```

File: keyboards/doio/kb38/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    [_QWERTY] = LAYOUT(
+        MO(1),            KC_BSPC,          RGB_RMOD, RGB_TOG, RGB_MOD,   KC_F1,     KC_F2,     KC_F3,
+        KC_NUM,  KC_PSLS, KC_PAST, KC_PMNS, KC_PAUS,  KC_SCRL, KC_PSCR, 
+        KC_P7,   KC_P8,   KC_P9,   KC_PPLS, KC_INS,   KC_HOME, KC_PGUP, 
+        KC_P4,   KC_P5,   KC_P6,            KC_END,   KC_DEL,  KC_PGDN, 
+        KC_P1,   KC_P2,   KC_P3,   KC_PENT,           KC_UP,
+        KC_P0,            KC_PDOT,          KC_LEFT,  KC_DOWN, KC_RGHT,  KC_HOME,    KC_END,    KC_B
+    ),
+    [_LAYERTWO] = LAYOUT(
+        KC_ESC,           KC_BSPC,          RGB_RMOD, RGB_TOG, RGB_MOD,   KC_A,   QK_REBOOT,   QK_BOOTLOADER,
```
Comment (waffle87) [inline keyboards/doio/kb38/keymaps/default/keymap.c:47]: Momentary layer change keys need a transparent key below them.
```suggestion
        _______,           KC_BSPC,          RGB_RMOD, RGB_TOG, RGB_MOD,   KC_A,   QK_REBOOT,   QK_BOOTLOADER,
```

File: keyboards/doio/kb38/keymaps/default/rules.mk [hunk 5]
```diff
@@ -0,0 +1 @@
+# forgot what goes here
```
Comment (waffle87) [inline keyboards/doio/kb38/keymaps/default/rules.mk:1]: Were you going to enable encoder map here? If not, this file should just be removed.

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb38/config.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+#define ENABLE_RGB_MATRIX_GRADIENT_UP_DOWN
+#define ENABLE_RGB_MATRIX_GRADIENT_LEFT_RIGHT
+#define ENABLE_RGB_MATRIX_BREATHING
+#define ENABLE_RGB_MATRIX_CYCLE_ALL
+#define ENABLE_RGB_MATRIX_RAINBOW_MOVING_CHEVRON
+#define RGB_MATRIX_DEFAULT_VAL RGB_MATRIX_MAXIMUM_BRIGHTNESS
+
+#define ENCODERS_PAD_A { C7, F7, D6 }
+#define ENCODERS_PAD_B { C6, E6, D7 }
+#define ENCODER_RESOLUTIONS { 4, 4, 2 }
```
Comment (drashna) [inline keyboards/doio/kb38/config.h:22]: Is there a particular reason for the setting?

Edit:  Ah, different sized knobs

File: keyboards/doio/kb38/kb38.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "quantum.h"
+
+//Display
+static void render_logo(void) {
```
Comment (drashna) [inline keyboards/doio/kb38/kb38.c:20]:
```suggestion
#ifdef OLED_ENABLE
static void render_logo(void) {
```

File: keyboards/doio/kb38/kb38.c [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    oled_write_P(qmk_logo, false);
+}
+
+bool oled_task_kb(void) {
+    if (!oled_task_user()) { return false; }
+    render_logo();
+    return true;
+}
+
+//Knobs
```
Comment (drashna) [inline keyboards/doio/kb38/kb38.c:36]:
```suggestion
#endif

//Knobs
#ifdef ENCODER_ENABLE
```

File: keyboards/doio/kb38/kb38.c [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    } else if (index == 2) {
+        // Volume Up/Down
+        if (clockwise) {
+            tap_code(KC_VOLU);
+        } else {
+            tap_code(KC_VOLD);
+        }
+    }
+    return true;
+}
```
Comment (drashna) [inline keyboards/doio/kb38/kb38.c:65]:
```suggestion
}
#endif
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb38/keymaps/default/rules.mk (hunk 5 above)
Comment (waffle87) [inline keyboards/doio/kb38/keymaps/default/rules.mk:1]: This file should be removed.

**Review by waffle87 — APPROVED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/doio/kb38/rgb_matrix_kb.inc [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        rgb_matrix_set_color(c, 0xFF, 0xFF, 0xFF);
+    }
+    for (d = 28; d < 33; d++) {
+        rgb_matrix_set_color(d, 0xFF, 0x00, 0x92);
+    }
+    for (e = 33; e < 38; e++) {
+        rgb_matrix_set_color(e, 0x00, 0x00, 0xFF);
+    }
+    for (bottom = 38; bottom < 44; bottom++) {
+        rgb_matrix_set_color(bottom, 0x00, 0x00, 0xFF);
```
Comment (fauxpark) [inline keyboards/doio/kb38/rgb_matrix_kb.inc:28]:
```suggestion
    for (int i = 0; i < 8; i++) {
        rgb_matrix_set_color(i, 0x00, 0x00, 0xFF);
    }
    for (int i = 8; i < 15; i++) {
        rgb_matrix_set_color(i, 0xFF, 0x00, 0x92);
    }
    for (int i = 15; i < 28; i++) {
        rgb_matrix_set_color(i, 0xFF, 0xFF, 0xFF);
    }
    for (int i = 28; i < 33; i++) {
        rgb_matrix_set_color(i, 0xFF, 0x00, 0x92);
    }
    for (int i = 33; i < 38; i++) {
        rgb_matrix_set_color(i, 0x00, 0x00, 0xFF);
    }
    for (int i = 38; i < 44; i++) {
        rgb_matrix_set_color(i, 0x00, 0x00, 0xFF);
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb38/keymaps/default/rules.mk (hunk 5 above)
Comment (fauxpark) [reply keyboards/doio/kb38/keymaps/default/rules.mk:1]: Still outstanding.

Comment (PepperKats) [reply keyboards/doio/kb38/keymaps/default/rules.mk]: > Still outstanding.

Sorry about that, forgot about that file. It's removed now.

**Pull request merged**